### PR TITLE
Visit ForeignItems when marking dead code

### PR DESCRIPTION
**Pull request by cjgillot**
Follow-up to #79318 

r? @lcnr

### CODE REVIEW

File: src/test/ui/lint/dead-code/type-in-foreign.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// check-pass
+
+#[repr(C)]
+struct Type(u8);
+
+extern "C" {
+    #[allow(dead_code)]
```
Comment (jyn514) [inline src/test/ui/lint/dead-code/type-in-foreign.rs:7]: Shouldn't this instead `warn(dead_code)` and check the error is correct?

Comment (cjgillot) [reply src/test/ui/lint/dead-code/type-in-foreign.rs:7]: I should add a comment: the test must not warn on `Type` line 4.

File: compiler/rustc_passes/src/dead.rs [hunk 2]
```diff
@@ -440,15 +422,31 @@ impl<'v, 'k, 'tcx> ItemLikeVisitor<'v> for LifeSeeder<'k, 'tcx> {
         }
     }
 
-    fn visit_trait_item(&mut self, _item: &hir::TraitItem<'_>) {
-        // ignore: we are handling this in `visit_item` above
+    fn visit_trait_item(&mut self, trait_item: &hir::TraitItem<'_>) {
```
Comment (lcnr) [inline compiler/rustc_passes/src/dead.rs:425]: is there a reason you did this move for trait items and not for impl items?

File: src/test/ui/lint/dead-code/type-in-foreign.rs (hunk 1 above)
Comment (lcnr) [reply src/test/ui/lint/dead-code/type-in-foreign.rs:7]: can you add `#[deny(dead_code)]` to this test?

**Review by lcnr — APPROVED** [review]

File: compiler/rustc_passes/src/dead.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_passes/src/dead.rs:425]: The treatment of impl items requires knowing whether they are implementing a trait. This information is not accessible from the impl item itself, so I could do the move.

File: compiler/rustc_passes/src/dead.rs [hunk 3]
```diff
@@ -440,15 +422,31 @@ impl<'v, 'k, 'tcx> ItemLikeVisitor<'v> for LifeSeeder<'k, 'tcx> {
...
     }
 
     fn visit_impl_item(&mut self, _item: &hir::ImplItem<'_>) {
         // ignore: we are handling this in `visit_item` above
     }
 
-    fn visit_foreign_item(&mut self, _item: &'v hir::ForeignItem<'v>) {}
+    fn visit_foreign_item(&mut self, foreign_item: &hir::ForeignItem<'_>) {
+        match foreign_item.kind {
+            hir::ForeignItemKind::Static(..) | hir::ForeignItemKind::Fn(..) => {
```
Comment (lcnr) [inline compiler/rustc_passes/src/dead.rs:444]: you need `has_allow_dead_code_or_lang_attr` here.

File: src/test/ui/lint/dead-code/lint-dead-code-3.rs [hunk 4]
```diff
@@ -57,9 +57,9 @@ mod blah {
     }
 }
 
-enum c_void {} //~ ERROR: enum is never used
+enum c_void {}
 extern {
-    fn free(p: *const c_void); //~ ERROR: function is never used
+    fn free(p: *const c_void);
```
Comment (lcnr) [inline src/test/ui/lint/dead-code/lint-dead-code-3.rs:62]: this should still warn, should it?

File: src/test/ui/lint/dead-code/type-in-foreign.rs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Verify that we do not warn on types that are used by foreign functions.
```
Comment (lcnr) [inline src/test/ui/lint/dead-code/type-in-foreign.rs:1]:
```suggestion
// Verify that we do not warn on types that are used by foreign functions
// which are considered used.
```

**Comment by lcnr** [comment]
please squash the commits

r=me afterwards

**Comment by cjgillot** [comment]
Squashed.

**Comment by lcnr** [comment]
Thanks :+1: 

@bors r+ rollup

**Pull request merged**